### PR TITLE
Use ReentrantLock instead of synchronized for JDK 21 compatibility

### PR DESCRIPTION
**Pull request by v-chernyshev**
**Feature type:**

Enhancement.

**Motivation:**

JDK 21 has introduced official support for virtual threads. It is a framework for efficient processing of blocking tasks in the simple thread-per-request style. Behind the scenes, the JVM runtime maintains a relatively small number of platform threads. A virtual thread is mounted onto a platform thread for execution. When the virtual thread blocks (e.g. on I/O) it is unmounted, and the platform thread is then free to execute a different task. See https://openjdk.org/jeps/444 for more details.

The current virtual threads implementation, however, does not play nicely with `synchronized` blocks and methods, especially when blocking operations are performed inside of them. In these situations the platform threads cannot be freed, and the JEP recommends using `ReentrantLock` instead. This PR replaces all usages of `synchronized` in `cdb2jdbc` with the recommended lock type.

**Current behaviour:**

The `cdb2jdbc` driver frequently causes pinning of the platform threads due to I/O under `synchronized` blocks (confirmed by running my service with the `-Djdk.tracePinnedThreads=full` flag).

**New behaviour:**

The pinning doesn't happen any longer as the driver uses the recommended synchronisation primitives.

Note:

https://github.com/bloomberg/comdb2/pull/4174 must be merged first to ensure that the build of the library succeeds.

### CODE REVIEW

File: cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/Comdb2Handle.java
Comment (v-chernyshev) [inline cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/Comdb2Handle.java:1537]: It's a very slight change in behaviour. Now `runStatement`, `next` and `close` use the same lock for the synchronisation. Previously `close` could run in parallel to the other two methods.

File: cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java
Comment (v-chernyshev) [inline cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java:48]: `TimeAndHosts` isn't used anywhere except this file, so it might as well be private.

File: cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java
Comment (v-chernyshev) [inline cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java:53]: I believe it's safer to use the map type that's specifically been designed with concurrency in mind. At the moment there is an unsynchronized call to `comdb2lcldb.remove` in `static void getDbHosts`.

**Comment by riverszhang89** [comment]
Thank you! I'll review this once the other pull request is merged.

File: cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java
Comment (v-chernyshev) [inline cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/DatabaseDiscovery.java:482]: As far as I can see, the removal of the lock is safe here:

- No other `hndl` operations are synchronized in this method.
- With `ConcurrentHashMap` `comdb2lcldb.get` is fully safe.
- The `TimeAndHosts` entries are never modified after they are created in `private static void queryDbHosts`.

File: cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/Comdb2Handle.java
Comment (v-chernyshev) [inline cdb2jdbc/src/main/java/com/bloomberg/comdb2/jdbc/Comdb2Handle.java:1642]: Let me know if you have an alternative naming suggestion for the private methods that are executed under their respective locks.

**Comment by v-chernyshev** [comment]
@riverszhang89 Please let me know if I need to do something with the PR to get the automated review from `roborivers`.

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>1/512 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>lostwrite</p>

**Review by riverszhang89 — APPROVED** [review]

**Comment by riverszhang89** [comment]
@v-chernyshev Thanks again for the pull request!!!

**Pull request merged**
